src/mpid/ch4/netmod/ofi/ofi_win.h

```diff
@@ -111,6 +111,249 @@ static inline int MPIDI_OFI_win_allgather(MPIR_Win * win, void *base, int disp_u
+        MPIDI_OFI_WIN(win).sep_tx_idx = -1;
+    }
+    goto fn_exit;
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_OFI_win_init_stx
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+static inline int MPIDI_OFI_win_init_stx(MPIR_Win * win)
+{
+    /* Activate per-window EP/counter using STX */
+    int ret, mpi_errno = MPI_SUCCESS;
+    struct fi_info *finfo;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_WIN_INIT_STX);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_WIN_INIT_STX);
+
+    finfo = fi_dupinfo(MPIDI_Global.prov_use);
+    MPIR_Assert(finfo);
+
+    /* Create shared context only in first call. */
+    if (MPIDI_Global.rma_stx_ctx == NULL) {
+        memset(finfo->tx_attr, 0, sizeof(struct fi_tx_attr));
+        /* A shared transmit context’s attributes must be a union of all associated
+         * endpoints' transmit capabilities. */
+        finfo->tx_attr->caps = FI_RMA | FI_WRITE | FI_READ | FI_ATOMIC;
+        finfo->tx_attr->msg_order = FI_ORDER_RAR | FI_ORDER_RAW | FI_ORDER_WAR | FI_ORDER_WAW;
+        finfo->tx_attr->op_flags = FI_DELIVERY_COMPLETE | FI_COMPLETION;
+        MPIDI_OFI_CALL_RETURN(fi_stx_context(MPIDI_Global.domain,
+                                             finfo->tx_attr,
+                                             &MPIDI_Global.stx_ctx, NULL /* context */), ret);
+        if (ret < 0) {
+            MPL_DBG_MSG(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                        "Failed to create shared TX context for RMA, "
+                        "falling back to global EP/counter scheme");
+            MPIDI_Global.rma_stx_ctx = NULL;
+            mpi_errno = MPI_ERR_OTHER;
+            goto fn_fail;
+        }
+    }
+
+    /* Set per window transmit attributes. */
+    MPIDI_OFI_set_rma_fi_info(win, finfo);
+    /* Still need to take out rx capabilities for shared context. */
+    finfo->rx_attr->caps = 0ULL;        /* RX capabilities not needed */
+
+    finfo->ep_attr->tx_ctx_cnt = FI_SHARED_CONTEXT;     /* Request a shared context */
+    finfo->ep_attr->rx_ctx_cnt = 0;     /* We don't need RX contexts */
+    MPIDI_OFI_CALL_RETURN(fi_endpoint(MPIDI_Global.domain,
+                                      finfo, &MPIDI_OFI_WIN(win).ep, NULL), ret);
+    if (ret < 0) {
+        MPL_DBG_MSG(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                    "Failed to create per-window EP using shared TX context, "
+                    "falling back to global EP/counter scheme");
+        mpi_errno = MPI_ERR_OTHER;
+        goto fn_fail;
+    }
+
+    if (MPIDI_OFI_win_set_per_win_sync(win) != MPI_SUCCESS) {
+        /* Close endpoint due to cmpl_cntr binding failure. */
+        MPIDI_OFI_CALL(fi_close(&MPIDI_OFI_WIN(win).ep->fid), epclose);
+        mpi_errno = MPI_ERR_OTHER;
+        goto fn_fail;
+    }
+
+    MPIDI_OFI_CALL(fi_ep_bind(MPIDI_OFI_WIN(win).ep, &MPIDI_Global.rma_stx_ctx->fid, 0), bind);
+    MPIDI_OFI_CALL(fi_ep_bind(MPIDI_OFI_WIN(win).ep,
```

Comment:
If `fi_ep_bind ` or `fi_enable` fails, the `MPIDI_OFI_WIN(win).ep` is overwritten in `MPIDI_OFI_win_init_global`, and will not be closed at `MPIDI_NM_mpi_win_free`.

Reply:
Okay, I will add error handling in `sxt_init` similar to `sep_init`.